File: public/methods/40_PORTFOLIO_METHODS/QM007_PORTFOLIO_BACKTESTING_REBALANCING/labs/r/qm007_hands_on.R
simulate_path <- function(df, rebalance_every) {
  target <- c(0.5, 0.5)
  w <- target
  wealth <- 1.0
  total_turnover <- 0.0
  for (k in seq_len(nrow(df))) {
    r <- c(df$asset_a_return[k], df$asset_b_return[k])
    rp <- sum(w * r)
    gross <- 1 + rp
    wealth <- wealth * gross
    drift <- w * (1 + r) / gross
    if (rebalance_every > 0 && df$period[k] %% rebalance_every == 0) {
      total_turnover <- total_turnover + 0.5 * sum(abs(target - drift))
      w <- target
    } else {
      w <- drift
    }
  }
  c(wealth = wealth, turnover = total_turnover)
}

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_path <- if (length(file_arg)) sub("^--file=", "", file_arg[1]) else normalizePath("qm007_hands_on.R")
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
df <- read.csv(file.path(root, "data", "qm007_two_asset_returns.csv"))

for (freq in c(1, 2, 0)) {
  label <- if (freq == 1) "every period" else if (freq == 2) "every 2 periods" else "buy-and-hold"
  ans <- simulate_path(df, freq)
  cat(sprintf("%s: terminal wealth=%.6f, cumulative one-way turnover=%.6f\n", label, ans["wealth"], ans["turnover"]))
}

File: public/methods/40_PORTFOLIO_METHODS/QM007_PORTFOLIO_BACKTESTING_REBALANCING/labs/python/qm007_hands_on.py
from pathlib import Path
import csv


def simulate(rows, rebalance_every):
    target = [0.5, 0.5]
    weights = target.copy()
    wealth = 1.0
    output = []
    for period, r_a, r_b in rows:
        rp = weights[0] * r_a + weights[1] * r_b
        gross = 1 + rp
        wealth *= gross
        drift = [weights[0] * (1 + r_a) / gross,
                 weights[1] * (1 + r_b) / gross]
        is_rebalance = rebalance_every > 0 and period % rebalance_every == 0
        if is_rebalance:
            turnover_one_way = 0.5 * sum(abs(t - d) for t, d in zip(target, drift))
            weights = target.copy()
        else:
            turnover_one_way = 0.0
            weights = drift
        output.append((period, rp, wealth, weights[0], turnover_one_way))
    return output


root = Path(__file__).resolve().parents[1]
rows = []
with (root / 'data/qm007_two_asset_returns.csv').open(newline='', encoding='utf-8') as f:
    for row in csv.DictReader(f):
        rows.append((int(row['period']), float(row['asset_a_return']), float(row['asset_b_return'])))

for frequency in (1, 2, 0):
    label = 'every period' if frequency == 1 else ('every 2 periods' if frequency == 2 else 'buy-and-hold')
    path = simulate(rows, frequency)
    total_one_way = sum(x[4] for x in path)
    print(f'{label:>15}: terminal wealth={path[-1][2]:.6f}, cumulative one-way turnover={total_one_way:.6f}')
